interfaces/builtin/uio.go

```diff
@@ -0,0 +1,103 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2019 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package builtin
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/interfaces/apparmor"
+	"github.com/snapcore/snapd/interfaces/udev"
+	"github.com/snapcore/snapd/snap"
+)
+
+// https://www.kernel.org/doc/html/latest/driver-api/uio-howto.html
+const uioSummary = `allows access to specific uio device`
+
+const uioBaseDeclarationSlots = `
+  uio:
+    allow-installation:
+      slot-snap-type:
+        - core
+        - gadget
+    deny-auto-connection: true
+`
+
+type uioInterface struct{}
+
+func (iface *uioInterface) Name() string {
+	return "uio"
+}
+
+func (iface *uioInterface) StaticInfo() interfaces.StaticInfo {
+	return interfaces.StaticInfo{
+		Summary:              uioSummary,
+		BaseDeclarationSlots: uioBaseDeclarationSlots,
+	}
+}
+
+var uioPattern = regexp.MustCompile(`^/dev/uio[0-9]+$`)
+
+func (iface *uioInterface) path(slotRef *interfaces.SlotRef, attrs interfaces.Attrer) (string, error) {
+	return verifySlotPathAttribute(slotRef, attrs, uioPattern, invalidDeviceNodeSlotPathErrFmt)
```

Comment:
to ease backporting and for clarity we probably want more specific errors:

invalidUioDeviceNodeSlotPathErrFmt = "slot %q path attribute must be a valid UIO device node"